### ISSUE TITLE
more generic `Pay` header format

### ISSUE THREAD

**Issue by michielbdejong** [interledger/rfcs#325]
Instead of `Pay: <destination> <amount> <shared-secret>`, let's use:

`Payment-Request: interledger <destination> <amount> <condition>` for one-of payment requests

and:

`Payment-Request: interledger-psk <destination> <amount> <shared-secret>` for psk-based http-ilp

**Comment by sharafian** [comment]
I think all of the interledger payment types should include application protocols. For instance, the first example you gave would be similar to IPR.

@adrianhopebailie is there anything in the web payments standards that describes a payment-request type header like this?

**Comment by michielbdejong** [comment]
It's indeed very similar to https://github.com/interledger/rfcs/blob/master/0011-interledger-payment-request/0011-interledger-payment-request.md#interledger-payment-request-format but I'm not sure if that's a reason to add that extra complication

**Comment by adrianhopebailie** [comment]
@sharafian no, but I have proposed something like this at IETF.

The only way to get anyone to look at our proposal will be to make it useful to people not doing ILP payments. Hence my suggestion to @michielbdejong of the format above.

**Comment by michielbdejong** [comment]
I thought about this again now, and maybe it would be good to say the value of the `Pay` header should be an IPR packet? Then we only need to define the IPR packet and not the other space-delimited formats.

**Comment by sharafian** [comment]
Well the thing about the current Pay header format is that the amount is just the cost of the current request. If the user wants to send more than that it's fine. If we switched from PSK to IPR then the user could no longer do that, and would have to make one payment per API call.

**Comment by michielbdejong** [comment]
Right, good point. I guess the best way to address this in the tutorials is to say we haven't standardized this yet, so the three different `Pay` headers used in the different tutorials are just examples of how we might define a machine-readable payment request in a http response header in the future

**Comment by adrianhopebailie** [comment]
Let's prioritize getting to consensus on this before we start changing the tutorials. Currently the letter shop tutorial makes not assertions about a standard. The bonus section simply illustrates the benefits of a machine readable mechanism for exchanging this data.

There are three reasons I proposed the format above to @michielbdejong, which I would summarize more specifically as follows:

`Pay: <payment-method> <destination> <amount> <payment-method-specific-data>`

1. Allowance for other payment methods

We are attempting to define a global standard for this header. We can't do that by trying to co-opt  HTTP and forcing anyone that uses this feature of the protocol to use ILP. (We can try but that brings with it a whole host of dependencies like first establishing ILP Addresses, our transport layer protocols etc as standards.)

It should be perfectly reasonable for someone to send a header like this:

`Pay: basic-card http://cardaquirer.com/api/endoint 10 <base64encodeddata>`

We may argue that is an inferior experience and we are probably right but let developers figure that out and prefer ILP-based protocols because they are better.

2. Alignment with Web Payments

This header format aligns nicely with the W3C Web Payments payment request model. The browser vendors have already shown unanimous support for this new standard so we should build on that not try to compete with it.

I have proposed a very rough Internet Draft for IETF in 2 weeks which I'll present to the HTTPbis and  and also the W3C Web Payments WG during our rechartering discussions at TPAC: https://datatracker.ietf.org/doc/draft-hope-bailie-http-payments/

This could use some work but I had to get it in before the deadline on Monday and gets me a few minutes on the agenda in the HTTPbis WG to at least solicit input on the idea from the WG that owns the HTTP standard.

3. The ability to support multiple transport layer protocols over ILP

We have focused mostly on PSK but there are use cases where IPR may be more appropriate and this also leaves the door open to introduce new versions of PSK or new transport layer protocols.

We could define a payment method called `interledger` which uses PSK as defined now by default and then have alternatives like `interledger-ipr` as they emerge.

One set of use cases where I think PSK may not be appropriate is in limited resource environments (IoT) where reducing the number of request/response round trips will be beneficial.
